plutus-chain-index/app/Marconi/Logging.hs

```diff
@@ -0,0 +1,122 @@
+      now <- getCurrentTime
+      modifyIORef' statsRef $ \stats ->
+        stats
+          { syncStatsAppliedBlocks = syncStatsAppliedBlocks stats + 1,
+            syncStatsLastAppliedBlocks = take numBlocks (now : syncStatsLastAppliedBlocks stats)
+          }
+      printMessage statsRef cp ct
+    update statsRef (RollBackward cp ct) = do
+      now <- getCurrentTime
+      modifyIORef' statsRef $ \stats -> stats {syncStatsLastRollback = Just now}
+      printMessage statsRef cp ct
+
+    printMessage statsRef cp ct = do
+      SyncStats {syncStatsAppliedBlocks, syncStatsLastAppliedBlocks, syncStatsLastRollback, syncStatsLastMessage} <- readIORef statsRef
+
+      let syncStatus = case (cp, ct) of
+            (_, ChainTipAtGenesis) ->
+              NotSynchronising
+            (ChainPointAtGenesis, _) ->
+              Synchronising 0
+            (ChainPoint (SlotNo chainPointSlot) _, ChainTip (SlotNo chainTipSlot) _header _blockNo)
+              -- TODO: MAGIC number here. Is there a better number?
+              -- 100 represents the number of slots before the
+              -- node where we consider the chain-index to be synced.
+              | chainTipSlot - chainPointSlot < 100 -> Synchronised
+            (ChainPoint (SlotNo chainPointSlot) _, ChainTip (SlotNo chainTipSlot) _header _blockNo) ->
+              let pct = (100 :: Double) * fromIntegral chainPointSlot / fromIntegral chainTipSlot
+               in Synchronising pct
+
+      let rate = case syncStatsLastAppliedBlocks of
+            _ : _ : _ ->
+              let timespan = diffUTCTime (List.head syncStatsLastAppliedBlocks) (List.last syncStatsLastAppliedBlocks)
+               in Just (fromIntegral numBlocks / realToFrac timespan :: Double)
+            _ ->
+              Nothing
+
+      let currentTipMsg = case syncStatus of
+            NotSynchronising -> ""
+            _                -> " Current point is " <> show cp <> "."
+
+      let rateMsg = case rate of
+            Nothing -> ""
+            Just r  -> " Current rate: " <> printf "%.2g" r <> " blocks/s."
+
+      let syncMsg = case syncStatus of
+            NotSynchronising -> " Not synchronising."
+            Synchronising p  -> printf " Synchornising (%0.2f%%)." p
+            Synchronised     -> " Synchronised."
+
+      let rollbackMsg = case syncStatsLastRollback of
+            Nothing -> ""
+            Just t  -> " Last rollback was on " <> show t <> "."
+
+      now <- getCurrentTime
+
+      let shouldPrint = case syncStatsLastMessage of
+            Nothing -> True
+            Just t
+              | diffUTCTime now t > minSecondsBetweenMsg -> True
+              | otherwise -> False
+
+      when shouldPrint $ do
+        modifyIORef' statsRef $ \stats -> stats {syncStatsLastMessage = Just now}
+        putStrLn $
+          "[" <> show now <> "]"
+            <> " Processed "
+            <> show syncStatsAppliedBlocks
```

Comment:
More of a discussion: You currently print the total number of processed blocks. Would it be better to print the number of processed blocks **since** the last log line?

Reply:
Yeah, this is actually a good simplification.